### PR TITLE
Fix GetFilter SQL for music artists, albums and songs nodes

### PR DESCRIPTION
**Pull request by DaveTBlake**
Fix a number of glitches and inconsistencies in the SQL created by `CMusicDatabase::GetFilter` for artists, albums and songs nodes. Some of them were introduced by my initial implementation of artist roles, and I would like to get it right before RC. Others are long standing but only show depending on the music data and the combination of multiple criteria. Also the better identification of individual artists, when there is more than one on a song, revealed some.

Unfortunately it is hard to see in what way the SQL has changed by just looking at the code, moreover without familiarity with the music db schema it is difficult to evaluate those changes.  Because of that I have delayed this PR for several months while I test it throughly (on both SQLite and MySQL).  I am confident that it is safe to release even at this stage of the cycle. 

Please don't be put off by the code diff! The SQL is constructed in a maintainable and modular way,  using string manipulation just as before, but the end result looks like a rewrite of this routine. In fact it may be better for me to provide a SQL diff instead if anyone is interested.

Some of the odd behaviour now fixed: 
- Inconsistent results between filtering albums by genre or filtering artists by genre and then navigating to albums, especially when album genre has been modified via NFO.
- Filtering by both genre _and_ artist showing albums with songs with that genre and songs with that artist, but not ensuring they were the same song.
- Displaying only album artists, but when navigate to albums for a selected artist also seeing those albums where they feature on some songs. It sometimes looks like random albums, and you get compilation albums which users complain about.
- Filtering artists for those that are album artists _and_ composers returning those that are an album artist but a composer of a song on some other album where they are not an artist. Similarly filtering albums for such an artist.
- Navigating from composers and seeing not only those albums with songs that they composed but also any album where they were album artist (and composed nothing).

### CODE REVIEW

**Comment by zag2me** [comment]
Tested and no issues that I can see with my music library, roles working well and totally awesome.

I can't comment too much on the individual cases, but no errors that I can see.

File: xbmc/dbwrappers/Database.cpp [hunk 1]
```diff
@@ -101,6 +101,55 @@ void CDatabase::Filter::AppendGroup(const std::string &strGroup)
...
+    join += " " + strJoin;
+}
+
+void CDatabase::ExistsSubQuery::AppendWhere(const std::string &strWhere, bool combineWithAnd /* = true */)
+{
+  if (strWhere.empty())
+    return;
+
+  if (where.empty())
+  where = strWhere;
```
Comment (notspiff) [inline xbmc/dbwrappers/Database.cpp]: indent

File: xbmc/dbwrappers/Database.cpp [hunk 2]
```diff
@@ -101,6 +101,55 @@ void CDatabase::Filter::AppendGroup(const std::string &strGroup)
...
+  {
+    where += combineWithAnd ? " AND " : " OR ";
+    where += strWhere;
+  }
+}
+
+bool CDatabase::ExistsSubQuery::BuildSQL(std::string & strSQL)
+{
+  std::string strWhere;
+  if (tablename.empty())
```
Comment (notspiff) [inline xbmc/dbwrappers/Database.cpp]: move decl after if

File: xbmc/dbwrappers/Database.cpp [hunk 3]
```diff
@@ -101,6 +101,55 @@ void CDatabase::Filter::AppendGroup(const std::string &strGroup)
...
+    strWhere += where;
+  }
+  if (!strWhere.empty())      
+    strSQL += " WHERE " + strWhere;
+
+  strSQL += ")";
+  return true;
+}
+
+
```
Comment (notspiff) [inline xbmc/dbwrappers/Database.cpp]: one line extra.

File: xbmc/music/MusicDatabase.cpp [hunk 4]
```diff
@@ -6342,104 +6342,130 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
...
+        ExistsSubQuery albumArtistSub("album_artist", "album_artist.idArtist = artistview.idArtist");
+        ExistsSubQuery songArtistSub("song_artist", "song_artist.idArtist = artistview.idArtist");
+        if (idRole > 0)
+          songArtistSub.AppendWhere(PrepareSQL("song_artist.idRole = %i", idRole));
+        if (idGenre > 0)
+        {
+          songArtistSub.AppendJoin("JOIN song_genre ON song_genre.idSong = song_artist.idSong");
+          songArtistSub.AppendWhere(PrepareSQL("song_genre.idGenre = %i", idGenre));
+        }
+        if (idRole <= 1 & idGenre > 0)
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: &&

File: xbmc/music/MusicDatabase.cpp [hunk 5]
```diff
@@ -6454,51 +6480,94 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
...
-        filter.AppendWhere(PrepareSQL(
-          "albumview.idAlbum IN (SELECT song.idAlbum FROM song JOIN song_artist ON song.idSong = song_artist.idSong "
-          "JOIN artist ON artist.idArtist = song_artist.idArtist "
-          "WHERE artist.strArtist like '%s' %s) OR " // All albums linked to this artist via songs
-          "albumview.idAlbum IN (SELECT album_artist.idAlbum FROM album_artist "
-          "JOIN artist ON artist.idArtist = album_artist.idArtist "
-          "WHERE artist.strArtist like '%s')", // All albums where album artists fit
-          option->second.asString().c_str(), strRoleSQL.c_str(), option->second.asString().c_str()));
-      // no artist given, so exclude any single albums (aka empty tagged albums)
+      if (idRole <= 1 & idGenre > 0)
```
Comment (notspiff) [inline xbmc/music/MusicDatabase.cpp]: &&

**Review by notspiff — COMMENTED** [review]
sorry for all the nits, just caught my attention.

**Comment by DaveTBlake** [comment]
>sorry for all the nits, just caught my attention.

Not at all, thanks for looking.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Any objections to this fix being merged?

In addition to my own extensive testing I have had positive feedback from users.  A test build, with this and the other as yet unmerged music PRs, is here http://mirrors.kodi.tv/test-builds/win32/KodiSetup-20161126-f6cce77-Music1126.exe

**Pull request merged**